src/write_gcm.c

```diff
@@ -3751,14 +3962,28 @@ static int wg_transmit_unique_segments(const wg_context_t *ctx,
       return -1;
     }
     DEBUG("write_gcm: next distinct segment has size %d", distinct_size);
-    int result = wg_transmit_unique_segment(ctx, queue, distinct_list);
-    if (result != 0) {
-      ERROR("write_gcm: wg_transmit_unique_segment failed.");
+    c_avl_tree_t *host_tree = wg_group_payloads_by_host(
+        distinct_list, &num_hosts);
+    if (num_hosts == -1) {
+      ERROR("write_gcm: wg_group_payloads_by_host failed.");
       wg_payload_destroy(distinct_list);
       wg_payload_destroy(residual_list);
       return -1;
     }
-    wg_payload_destroy(distinct_list);
+    c_avl_iterator_t *iterator = c_avl_get_iterator(host_tree);
+    while (c_avl_iterator_next(iterator, (void **) &key,
+              (void **) &hook) == 0) {
+      DEBUG("write_gcm: Dispatching payloads for host %s", key);
+      wg_payload_t *list = ((wg_payload_hook_t *) hook)->payload;
+      int result = wg_transmit_unique_segment(ctx, queue, list);
+      if (result == -2) {
+        WARNING("write_gcm: Dropped payload with hostname: %s", key);
+      } else if (result != 0) {
+        ERROR("write_gcm: wg_transmit_unique_segment failed.");
+      }
+    }
+    c_avl_iterator_destroy(iterator);
+    wg_destroy_host_tree(host_tree);
```

Comment:
*[Optional]* Should we not be able to do free the appropriate pieces right after `wg_transmit_unique_segment` in the previous loop, rather than looping over the whole tree again?